### PR TITLE
Support decorated constructors

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/1706
Fixes https://github.com/python/mypy/issues/5398

Previously, decorated constructors either resulted in silent `Any` types for class objects (in case of a decorated `__init__()`) or incorrect signatures (in case of a decorated `__new__()`).

This PR adds support for decorated constructors, implementation is straightforward.

This PR supersedes https://github.com/python/mypy/pull/5798.

### CODE REVIEW

**Comment by gvanrossum** [comment]
I'm sorry, I don't think I can review this.

File: mypy/checkmember.py [hunk 1]
```diff
@@ -797,17 +804,35 @@ def type_object_type(info: TypeInfo, builtin_type: Callable[[str], Instance]) ->
...
+        t = method.type
+    signature = bind_self(t)
+    return type_object_type_from_function(signature, info, method.info, fallback)
+
 
+def is_valid_constructor(n: Optional[SymbolNode]) -> bool:
+    """Does this node represents a valid constructor method?
 
-def type_object_type_from_function(init_or_new: FuncBase,
+    This includes nomral functions, overloaded functions, and decorators
```
Comment (msullivan) [inline mypy/checkmember.py]: typo - "nomral"

File: mypy/checkmember.py [hunk 2]
```diff
@@ -797,17 +804,35 @@ def type_object_type(info: TypeInfo, builtin_type: Callable[[str], Instance]) ->
...
+        method = init_method.node
     # Construct callable type based on signature of __init__. Adjust
     # return type and insert type arguments.
-    return type_object_type_from_function(method, info, fallback)
+    if isinstance(method, FuncBase):
+        t = function_type(method, fallback)
+    else:
+        assert isinstance(method.type, FunctionLike)  # is_valid_constructor() ensures this
+        t = method.type
+    signature = bind_self(t)
```
Comment (msullivan) [inline mypy/checkmember.py]: Any particular reason for moving this `bind_self` out of `type_object_type_from_function`? I don't have any objection, I'm just wondering if there is a technical reason I'm missing or you just like it more.

Comment (ilevkivskyi) [reply mypy/checkmember.py]: No particular reason. I think I started refactoring by pulling `function_type()` out of there, and pulled whole line out of there. But looking again at it I think it makes sense to keep bind and map together, so I will put it back.

**Comment by ilevkivskyi** [comment]
> How much damage does it cause for internal codebases?

Not so much actually, around 60 errors S+C.

**Pull request merged**